justStaticExecutables: Forbid references to GHC

9999years opened the pull request:
## Description of changes

This makes `justStaticExecutables` error if the produced store path contains references to GHC. This is almost always erroneous and due to the generated `Paths_*` module being imported. This helps prevent `justStaticExecutables` from producing binaries with closure sizes in the gigabytes.

See: https://github.com/NixOS/nixpkgs/issues/164630

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

9999years committed:
haskell.lib.compose.justStaticExecutables: Forbid references to GHC

This makes `justStaticExecutables` error if the produced store path
contains references to GHC. This is almost always erroneous and due to
the generated `Paths_*` module being imported. This helps prevent
`justStaticExecutables` from producing binaries with closure sizes in
the gigabytes.

See: https://github.com/NixOS/nixpkgs/issues/164630

Co-authored-by: sternenseemann <sternenseemann@systemli.org>

maralorn commented:
I like this.

A bit sad that this makes justStaticExecutables better than enableSeparateBinOutput. I forgot whether there is a per output option regarding requisites.

We will need to make sure that this doesn’t break any of the Haskell binaries that we ship in nixpkgs so this definitely needs to merge into haskell-updates.

9999years commented:
@maralorn Agreed, how do I do that? Just change the merge target?

sternenseemann reviewed (changes requested):
`disallowedRequisites` needs to become an argument to `haskellPackages.mkDerivation` and transparently passed on to `stdenv.mkDerivation` in there. Otherwise `disallowedRequisites` will only be set depending on override order, i.e. any `overrideCabal`-based function applied after `justStaticExecutables` would revert the change again, as it uses a different fix point from `overrideAttrs`.

9999years commented:
@sternenseemann I'm not sure how to make that work; I can add `disallowedRequisites` to the generic builder, but then I don't have access to the `passthru.compiler` attribute. I can add a `ghc` argument with a default value, but then I can't access that in the override unless it's changed from the default due to https://github.com/NixOS/nix/issues/334

sternenseemann commented:
@9999years `passthru.compiler` is set in that function, so it is available as the `ghc` argument passed in.

sternenseemann commented:
Ah, I get what you mean. In principle you can literally add an `disallowGhcReference` argument tbh.

sternenseemann commented:
Or `drv: let old = drv; in overrideCabal { disallowedRequisites = [ old.passthru.compiler ]; } old` should also work?

9999years commented:
OK, I took a shot at it, let me know what you think.

I also checked that the docs render correctly:

<img width="808" alt="image" src="https://github.com/NixOS/nixpkgs/assets/15312184/92ad6a14-0124-426e-877e-3161fcb51487">

9999years commented:
@sternenseemann Still looking for a review on this!

sternenseemann commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -20,6 +21,11 @@ in
+  #
+  # `haskell.lib.justStaticExecutables` will add this to `disallowedRequisites`
+  # unless `needsGHCAtRuntime` is set.
+, ghc ? originalGHC
```
Why are you threading it through in this way? It doesn't really make sense that you can freely pass `ghc = …` to `haskellPackages.mkDerivation` as that already has a predetermined GHC and no other would work. You can simply rely on `ghc` as it is already in scope, can't you?

9999years replied:
Hmm. I guess you're right that this is usually redundant, but I see no reason why it wouldn't work. I'll remove it for now.

sternenseemann commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -20,6 +21,11 @@ in
+  # The GHC to build this package with.
+  #
+  # `haskell.lib.justStaticExecutables` will add this to `disallowedRequisites`
+  # unless `needsGHCAtRuntime` is set.
```
(I think the argument name is outdated here.)

sternenseemann reviewed (commented):
Sorry for the wait!

9999years commented:
PTAL

maralorn commented:
This looks reasonable and simple. Really appreciate the documentation work! Thank you.

sternenseemann reviewed (commented):
This looks good to me! Last bit of busywork would be to add an entry to the changelog noting the change so people upgrading aren't confused if their derivations should start to fail.

9999years commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -645,9 +667,17 @@ stdenv.mkDerivation ({
   passthru = passthru // rec {
 
-    inherit pname version;
+    inherit pname version disallowGhcReference;
```
So that other derivations or Nix code can introspect it. It might be useful, especially for debugging.

9999years commented:
Sure, where does the changelog entry go?

sternenseemann commented:
> Sure, where does the changelog entry go?

`nixos/doc/manual/release-notes/rl-2405.section.md`, in the breaking changes section somewhere.

9999years commented:
Thanks, added a note to the breaking changes section. Let me know how it looks!

sternenseemann committed:
haskellPackages.mkDerivation: no rebuild w/o disallowedRequisites

This change ensures that packages won't be rebuild compared to before
the introduction of disallowedRequisites/disallowGhcReference unless
they use one of those arguments.

It may be nice to revert this in the future (via staging) for greater
simplicity, but will help with initial regression testing.

sternenseemann commented:
I've realized that we can't backport this anymore since this is technically a breaking change. I've moved the changelog entry to the 24.11 release notes as a result. We *could* backport a version of this which adds the new arguments to `haskellPackages.mkDerivation`, but doesn't change the behavior of `justStaticExecutables`.

I've also noticed that we haven't documented `disallowedRequisites` and `disallowedGhcReference` to the `mkDerivation` documentation in the manual, but maybe that's okay since `disallowedRequisites` is undocumented for `stdenv.mkDerivation` as well…

sternenseemann commented:
@9999years Evals newer than 1806586 of [haskell-updates](https://hydra.nixos.org/jobset/nixpkgs/haskell-updates) contain your change. There are already a few regressions I've seen on aarch64-darwin. This is not really surprising as GHC can't properly eliminate dead code on that platform. It may be prudent to ignore `disallowGhcReference` on that platform.

9999years commented:
> There are already a few regressions I've seen on aarch64-darwin

Which attributes are impacted?

sternenseemann commented:
@9999years They should accumulate [here](https://hydra.nixos.org/eval/1806586#tabs-now-fail). Note that a lot of builds are still pending so the list will get longer as time goes on.
